### PR TITLE
player list is now only updated at the end of the login

### PR DESCRIPTION
**Pull request by optimumtact**
Client is removed from the player list is login fails due to an exception

If a runtime occurs during the login proc because the client has gone
again, then logout is never called and it seems like this leads to nulls
in the player list

We catch runtimes, remove the mob from the player list and then
rethrow

https://tgstation13.org/parsed-logs/sybil/data/logs/2019/03/18/round-104032/runtime.log

will show this pattern quite clearly, when it's available publically

### CODE REVIEW

File: code/modules/mob/login.dm [hunk 1]
```diff
@@ -54,3 +53,7 @@
 		log_played_names(client.ckey,name,real_name)
 
 	log_message("Client [key_name(src)] has taken ownership of mob [src]([src.type])", LOG_OWNERSHIP)
+
+	//If login fails and there is no client or src, we want to bail, because otherwise there will be no logout
+	if(client)
```
Comment (optimumtact) [inline code/modules/mob/login.dm]: I left out src, but I'm almost tempted to add it, even if it seems cargo cult

**Comment by AnturK** [comment]
Starting at least with world status update things depend on the player list change.

**Comment by optimumtact** [comment]
@AnturK not getting what you're leaning at?

**Comment by optimumtact** [comment]
ah I see what you mean, I'll audit them a bit later, but they can be moved to after it, or better yet, a post login callback

**Comment by MrStonedOne** [comment]
The cause of the logout failure is the mob getting hard deleted while it has a client connected to it.

This does not call logout.

File: code/modules/mob/login.dm (hunk 1 above)
Comment (SpaceManiac) [reply code/modules/mob/login.dm]: I'm pretty sure the semantics are that if something is hard deleted, any sleeping procs with it as `src` are cancelled

**Comment by optimumtact** [comment]
Changed this to catch exceptions remove the mob from the player list and then rethrow

should work just as robustly and doesn't require me to rewrite the login code

**Comment by MrStonedOne** [comment]
Exception catching extends down the stack, making it so an exception that would normally just crash a small 3 line proc 7 calls down the chain now crashes the entire chain.

Doing it that way will likely have unintended side effects as well as significantly increase the likelihood that mob changes and assignments will fail oddly.

If you want to catch exceptions within a single proc without it extending down the call chain, you wrap a proc around it, and always return `TRUE` on all control paths, now you can detect runtimes by checking the return value. This is how the MC does it.

**Comment by ShizCalev** [comment]
conflict

**Comment by SpaceManiac** [comment]
Would love to see this fixed but it looks like this will not do it

**Pull request closed without merging**